cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherClientHandler.java

```diff
@@ -0,0 +1,115 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tether;
+
+import com.google.gson.Gson;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.namespace.NamespaceAdmin;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpResponse;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * {@link io.cdap.http.HttpHandler} to manage tethering client v3 REST APIs
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+public class TetherClientHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TetherClientHandler.class);
+  private static final Gson GSON = new Gson();
+  static final String CREATE_TETHER = "/v3/tethering/connect";
+  // Following constants are same as defined in KubeMasterEnvironment
+  static final String NAMESPACE_CPU_LIMIT_PROPERTY = "k8s.namespace.cpu.limits";
+  static final String NAMESPACE_MEMORY_LIMIT_PROPERTY = "k8s.namespace.memory.limits";
+
+  private final TetherStore store;
+  private final NamespaceAdmin namespaceAdmin;
+  private String instanceName;
+
+  @Inject
+  TetherClientHandler(CConfiguration cConf, TetherStore store, NamespaceAdmin namespaceAdmin) {
+    this.store = store;
+    this.namespaceAdmin = namespaceAdmin;
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+  }
+
+  /**
+   * Initiates tethering with the server.
+   */
+  @POST
+  @Path("/tethering/create")
+  public void createTether(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherCreationRequest tetherCreationRequest = GSON.fromJson(content, TetherCreationRequest.class);
+
+    List<PeerInfo> peers = store.getPeers();
+    Optional<PeerInfo> peer = peers.stream()
+      .filter(p -> p.getName().equals(tetherCreationRequest.getInstance()))
+      .findFirst();
+    if (peer.isPresent()) {
+      LOG.info("Peer {} is already present in state {}, ignoring tethering request",
```

Comment:
If the state is rejected, can we still add same peer again? Or that has to be done on server side?

Reply:
Rejected is a terminal state. Once tethering is rejected, it can only be deleted. Peer can be re-added after deleting on both sides.

Reply:
I see